src/lx/print/json.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2019 Jamey Sharp
+ *
+ * See LICENCE for the full copyright terms.
+ */
+
+#include <assert.h>
+#include <stdio.h>
+#include <errno.h>
+
+#include <adt/set.h>
+
+#include <fsm/fsm.h>
+#include <fsm/pred.h>
+
+#include "libfsm/internal.h" /* XXX */
+
+#include "lx/ast.h"
+#include "lx/print.h"
+
+/* TODO: centralise with libfsm */
+static unsigned int
+indexof(const struct fsm *fsm, const struct fsm_state *state)
+{
+	struct fsm_state *s;
+	unsigned int i;
+
+	assert(fsm != NULL);
+	assert(state != NULL);
+
+	for (s = fsm->sl, i = 0; s != NULL; s = s->next, i++) {
+		if (s == state) {
+			return i;
+		}
+	}
+
+	assert(!"unreached");
+	return 0;
+}
+
+static unsigned int
+zindexof(const struct ast *ast, const struct ast_zone *zone)
+{
+	struct ast_zone *z;
+	unsigned int i;
+
+	assert(ast != NULL);
+	assert(zone != NULL);
+
+	for (z = ast->zl, i = 0; z != NULL; z = z->next, i++) {
+		if (z == zone) {
+			return i;
+		}
+	}
+
+	assert(!"unreached");
+	return 0;
+}
+
+static int
+print_zone(FILE *f, const struct ast *ast, const struct ast_zone *z)
+{
+	struct fsm_state *s, *st;
+	int ret;
+
+	assert(f != NULL);
+	assert(z != NULL);
+	assert(z->fsm != NULL);
+	assert(fsm_all(z->fsm, fsm_isdfa));
+	assert(ast != NULL);
+
+	fprintf(f, "    {\n");
+	fprintf(f, "      \"initial_state\": %u,\n", indexof(z->fsm, z->fsm->start));
+	fprintf(f, "      \"states\": [\n");
+
+	for (s = z->fsm->sl; s != NULL; s = s->next) {
+		struct fsm_edge *e;
+		struct set_iter it;
+
+		fprintf(f, "        {\n");
+
+		e = set_first(s->edges, &it);
+
+		if (fsm_isend(z->fsm, s)) {
+			const struct ast_mapping *m = s->opaque;
+			assert(m != NULL);
+
+			if (m->token != NULL) {
+				fprintf(f, "          \"token\": \"$%s\",\n", m->token->s);
+			}
+			if (m->to != NULL) {
+				fprintf(f, "          \"next_zone\": %u,\n", zindexof(ast, m->to));
+			}
+
+			fprintf(f, "          \"accepts\": true%s\n", e != NULL ? "," : "");
+		}
+
+		if (e != NULL) {
+			int first = 1;
+			fprintf(f, "          \"transitions\": [");
+			for (; e != NULL; e = set_next(&it)) {
+				struct set_iter jt;
+				for (st = set_first(e->sl, &jt); st != NULL; st = set_next(&jt)) {
```

Comment:
libfsm has an IR for code generation which provides a few optimisations for generated DFA. I'm not sure if you'd want to use it here. It provides different "strategies" for each state, where e.g. symbols are presented grouped as `x..y` ranges, or e.g. if all symbols transition to the same state. If you're going on to generate code from the json, I think it probably would make sense to make use of that.

To use that IR, I'd first add json output to libfsm (cut & pasted from the current C), which just walks through that IR and outputs what it's given. The IR was introduced here: https://github.com/katef/libfsm/pull/94

Then `re -pl json xyz` would give json for a single regexp, too.